### ISSUE TITLE
Once "context deadline exceeded" error happened, need to restart service.

### ISSUE THREAD

**Issue by shawn111** [mattermost/mattermost-server#21641]
#### Summary
When the error happened, can not access mattersmost service.
It would redirect to login page and show "Token Error", like https://forum.gitlab.com/t/mattermost-token-error/65279/2.

Check mattermost.log, once this error happen, all db connection were broken.
Nov 11 09:41:13 messaging-server mattermost: {"timestamp":"2022-11-11 09:41:13.085 +08:00","level":"error","msg":"Failed to list plugin key values","caller":"app/plugin_key_value_store.go:167","page":0,"perPage":1000,"error":"failed to get PluginKeyValues with pluginId=playbooks: context deadline exceeded"}

Not sure it is related to https://github.com/jackc/pgx/issues/1313.

Bug report in one concise sentence

#### Steps to reproduce
Mattermost Version: 7.1.3
Database Schema Version: 89
Database: postgres (13.8)

A complete restart of the system solves the problem, but it comes back every 4-5 day.
We check db / memory / network status, can not find the problem.

#### Expected behavior
Even the error happened, it should broken that query not all of queries after that.

#### Observed behavior (that appears unintentional) 
Once "context deadline exceeded" error happened, need to restart service.

#### Possible fixes
Not sure, should we upgrade pgx to fix it?
https://github.com/jackc/pgx/issues/1313

**Comment by amyblais** [comment]
Opened: https://mattermost.atlassian.net/browse/MM-48353.

**Comment by agnivade** [comment]
Hello @shawn111 - an error of "context deadline exceeded" means that the database exceeded the query timeout set by the application. 

I'd suggest to check the database logs for any issues. But as you mentioned above, the db does not appear to have any issues.

I think I'd need some more information about your setup to be able to provide further info.
- Is it a docker setup?
- Are both the application and DB in the same host?
- What are your system specs and what is your DB size?

Let me tag @anx-ag who's had some experience debugging these issues before. Alexander, do you know what could be the issue here?

**Comment by anx-ag** [comment]
If it happens unregularly or only when the system has been running for some days, there might be an autovacuum blocking things or reducing performance so that the other queries cannot finish. It's also possible that the application's connection pools get full because of that or that you reach max connections on your database server and the application therefore stops working for new queries while the older ones are timing out.

Additional to what @agnivade requested, can you please also share the `SqlSettings` block of your `config.json` file (with the passwords and other sensitive details removed, of course)?
When you say "We check DB": what exactly has been checked? Did you also look at the running processes in the database when the problems are ongoing?

The output of the following query while you're experiencing the problem would be very helpful:

```sql
select * from pg_stat_activity;
```

**Comment by shawn111** [comment]
@agnivade yes, as beginning. I also thought this issue is caused by db loading.
However, when I check the grafana postgres dashboard. I thought the problem might cause from mattermost side.

{"timestamp":"2022-11-13 09:14:44.587 +08:00","level":"error","msg":"GetChannelMembersForUserWithPagination: Unable to get the channel members., failed to find ChannelMembers data with and userId=1gigwsncqibwdp5euyp64xy3fw: context deadline exceeded","caller":"app/plugin_api.go:940","plugin_id":"com.mattermost.calls","origin":"main.(*Plugin).handleGetAllChannels api.go:129"}

In order to fix it, we restarted the mattermost service
![image](https://user-images.githubusercontent.com/346761/201641121-0e0ae027-afb3-48cd-8a4b-f33ce4d7ead5.png)

@anx-ag , Here is the latest 4 days info.
![image](https://user-images.githubusercontent.com/346761/201642011-ba47cf34-22d3-4443-ba0b-b2a1c518e526.png)
![image](https://user-images.githubusercontent.com/346761/201642339-76329d51-5f9d-4fc7-91a5-ceda001d27a1.png)

About my env, mattermost run direct in host server, but postgres already move into k8s env.
They are not in the same internal lan.

It might be the root cause of "context deadline exceeded", but even this issue happened, it should be covered after a moment.

**Comment by shawn111** [comment]
![image](https://user-images.githubusercontent.com/346761/201645944-e5b48c1b-1764-41bd-a288-2fa851d78aed.png)

BTW, I'm curious about it seem there is a big query every hours, even in mid night.
We should not have lots of users in mid night.

**Comment by shawn111** [comment]
```
pg_dump: error: query failed: ERROR:  permission denied for schema zhparser
pg_dump: error: query was: LOCK TABLE zhparser.zhprs_custom_word IN ACCESS SHARE MODE
```
@agnivade It seem it might cause by my setting.
We use an zhparser extension, maybe the "context deadline exceeded" is related with the LOCK.

**Comment by anx-ag** [comment]
Is the zhparser extension enabled in the mattermost database? I must admit I was unable to follow the documentation on the Github page properly, but the hourly peaks are most likely not caused by this extension. If you want to find out what queries these are, you could try to enable query logging a few minutes before this happens and disable query logging again afterwards, so we can find out what causes these spikes but they're most likely unrelated to your problem.

Also did you find time to run the query I sent you and did what was the result? Please do also send the requested configuration settings of your config.json.
Are there any other statistics or metrics you have from your database server? The screens you shared here are not really helpful when debugging issues like these, other metrics like query duration, system load, number of queries/second, etc. would be more interesting.

**Comment by shawn111** [comment]
@anx-ag It's might be a network issue between my pg server and mattermost server.

Ping time is not stable.
```
2022-11-29 08:10:04 64 bytes from 172.16.0.65: icmp_seq=61239 ttl=62 time=0.327 ms
2022-11-29 08:10:05 64 bytes from 172.16.0.65: icmp_seq=61240 ttl=62 time=2.09 ms
2022-11-29 08:10:06 64 bytes from 172.16.0.65: icmp_seq=61241 ttl=62 time=0.430 ms
2022-11-29 08:10:07 64 bytes from 172.16.0.65: icmp_seq=61242 ttl=62 time=0.378 ms
```

Database loading still far to reach the limitation. 
It's the reason why "context deadline exceeded" error happen.

However, it point out once "context deadline exceeded" happened, all db connection would be broken.
I'm not sure is it related to https://github.com/jackc/pgx/issues/1313

**Comment by anx-ag** [comment]
Getting `Context deadline exceeded` messages is just the result, not the cause. This usually gets printed when the queries take longer than allowed and long running queries lock your available database connections which might then result in a situation where no free slots are available.

In order to debug this further I would really need the requested outputs and answers to my questions. Do you need help with running those queries or should I rephrase the questions if you've trouble understanding them? Just let me know.

**Comment by shawn111** [comment]
Got it, I thought in my env is quite clear about network issue to cause this issue.
Thank for your help.
We need fix the network issue first.
